### DISCUSSION TITLE
Block model change if block on top

### DISCUSSION THREAD

**Discussion by tenDeeZed** [FabricMC/fabric-api#1924] [Mod Dev Support] [answered]
Hello there. I want to create something but I'm not sure how to do it. Can someone help me please?

I want to create block that change it's model and outline shape when you place another (any) block on top of it and change back to default model when you destroy block on top.  I have prepared everything but I'm not sure how to setup the class for this block, more specifically I don't know how to make it change...

I'm kinda new and I'm learning by reading MC code. I found something similar in WallBlock (WallPost) but I'm kinda confused from it. And anything else like Farmland or DirtPath is done by replacing block...

**Comment by warjort** [comment]
> And anything else like Farmland or DirtPath is done by replacing block...

That is the usual way it is done. 

The basic block models use the block id or block states to select which one to use. So you listen to neighbouring block updates and either 
* change your block id (not very subtle)
* or give it a different state like "blockontop=true" which lets you use a different variant/model in the blockstates json

The other more complicated way is to use a block entity which can have any NBT values. 
Then you can write your own custom rendering based on the NBT values.

**Comment by warjort** [comment]
> I found something similar in WallBlock (WallPost) but I'm kinda confused from it

This is using the built in infrastructure with some extra stuff on top to handle calculating where it is connected to other blocks.

The logic starts at ````AbstractBlockState.updateNeighbors```` which iterates over all 6 directions to ask the neighbouring blocks if they want to change state when this block changes.
The WallBlock responds by looking around itself to see what blocks it should be connected to and calculating a shape/state.
It then returns that state for the updateNeighbors to actually do the modification - calling back into Block.replace().

If you look at say ````assets/minecraft/blockstates/cobblestonewall.json````
you will see the states getting referenced, e.g. "up", "east"
But there it is using a multipart model to render multiple models from the relevant states.
You probably just want a variant model like the stairs use.
````
  "variants": {
    "": {
      "model": "default model here"
    },
    "blockontop=true": {
      "model": "changed model name here"
    },
etc.
````

**Reply by tenDeeZed** [reply]
Thanks for reply. Can you send me example of class? I tried something like this
`public static final BooleanProperty TALL = BooleanProperty.of("tall");`

`public MyBlock(Settings settings) {
        super(settings);
        setDefaultState(getStateManager().getDefaultState().with(TALL, false));
    }`
    
    public BlockState getStateForNeighborUpdate(BlockState state, Direction direction, BlockState neighborState, WorldAccess world, BlockPos pos, BlockPos neighborPos) {
        if (direction == Direction.UP) {
            world.setBlockState(pos, state.with(TALL, true));
        }
    }

But it seems to be completely wrong :D

**Comment by warjort** [comment, marked as answer]
> But it seems to be completely wrong :D

You shouldn't be setting the block state directly. You return it to the caller.

The caller will:
* check if the state has actually changed which avoids unnecessary updates and even potentially infinite loops
* set the update flags to do the correct processing/notifications
* other book keeping stuff (like dropping the block item if you change the block state to "air")

If you want a simple example, look at how the repeater block sets its (un)locked state.
